### PR TITLE
QSERVER-17813 GraphQL Field Duplication Enabled

### PR DESCRIPTION
**Pull request by Vishal-Rao-SU**

### CODE REVIEW

**Review by AEkon — APPROVED** [review]

File: qudini-reactive-graphql/src/main/java/com/qudini/reactive/graphql/http/GraphQLHandler.java [hunk 1]
```diff
@@ -28,6 +29,9 @@ public final class GraphQLHandler {
     private final GraphQLSchema schema;
     private final DataFetcherExceptionHandler exceptionHandler;
 
+    @Value("${qudini-reactive.graphql-max-depth}")
```
Comment (jmaister) [inline qudini-reactive-graphql/src/main/java/com/qudini/reactive/graphql/http/GraphQLHandler.java:32]: Please document this parameter and how it affects the queries in the README

File: qudini-reactive-graphql/src/main/java/com/qudini/reactive/graphql/http/GraphQLHandler.java [hunk 2]
```diff
@@ -42,7 +46,7 @@ private Mono<Map<String, Object>> execute(ContextView context, GraphQLRequest re
         var input = request.toExecutionInput(context, registry);
         var graphql = GraphQL
                 .newGraphQL(schema)
-                .instrumentation(new DataLoaderDispatcherInstrumentation())
+                .instrumentation(new MaxQueryDepthInstrumentation(maxDepth))
```
Comment (jmaister) [inline qudini-reactive-graphql/src/main/java/com/qudini/reactive/graphql/http/GraphQLHandler.java:49]: Is the new class equivalent in terms of functionality? Is it going to return the same results?

Comment (jmaister) [inline qudini-reactive-graphql/src/main/java/com/qudini/reactive/graphql/http/GraphQLHandler.java:49]: Maxdepth could be null here.

File: qudini-reactive-graphql/src/main/java/com/qudini/reactive/graphql/http/GraphQLHandler.java [hunk 3]
```diff
@@ -42,7 +46,16 @@ private Mono<Map<String, Object>> execute(ContextView context, GraphQLRequest re
         var input = request.toExecutionInput(context, registry);
         var graphql = GraphQL
                 .newGraphQL(schema)
-                .instrumentation(new DataLoaderDispatcherInstrumentation())
+                .instrumentation(new MaxQueryDepthInstrumentation(maxDepth, queryDepthInfo -> {
+                    if (queryDepthInfo.getDepth() > maxDepth) {
+                        try {
+                            throw new Exception("Query depth exceeded maximum allowed depth of" + maxDepth);
```
Comment (jmaister) [inline qudini-reactive-graphql/src/main/java/com/qudini/reactive/graphql/http/GraphQLHandler.java:52]: Exception thrown, catched and thrown again. This is not right.

It should me something like this:

```java
.instrumentation(new MaxQueryDepthInstrumentation(maxDepth, queryDepthInfo -> {
    if (queryDepthInfo.getDepth() > maxDepth) {
        throw new MaxQueryDepthExceededException("Query depth exceeded maximum allowed depth of " + maxDepth);
    }
    return false;
}))

File: qudini-reactive-graphql/src/main/java/com/qudini/reactive/graphql/http/GraphQLHandler.java (hunk 2 above)
Comment (Vishal-Rao-SU) [reply qudini-reactive-graphql/src/main/java/com/qudini/reactive/graphql/http/GraphQLHandler.java:49]: wrt latest reference 
Exception is thrown if Query Depth is exceeded

{
  "errors": [
    {
      "message": "maximum query depth exceeded 4 > 3",
      "extensions": {
        "classification": "ExecutionAborted"
      }
    }
  ]
}

File: qudini-reactive-graphql/src/main/java/com/qudini/reactive/graphql/http/GraphQLHandler.java [hunk 4]
```diff
@@ -28,6 +29,9 @@ public final class GraphQLHandler {
     private final GraphQLSchema schema;
     private final DataFetcherExceptionHandler exceptionHandler;
 
+    @Value("${qudini-reactive.graphql-max-depth:15}")
```
Comment (jmaister) [inline qudini-reactive-graphql/src/main/java/com/qudini/reactive/graphql/http/GraphQLHandler.java:32]: Let's use a lower and safer default here. Like 5.
Then projects can increase it if needed.

Comment (Vishal-Rao-SU) [reply qudini-reactive-graphql/src/main/java/com/qudini/reactive/graphql/http/GraphQLHandler.java]: taken care

**Review by jmaister — APPROVED** [review]

**Review by ajpachoriya — APPROVED** [review]

**Comment by ajpachoriya** [comment]
Please address the comment above from Jordi about the exception handling, if taken care kindly mark resolved.

**Review by AEkon — DISMISSED** [review]

**Review by ajpachoriya — DISMISSED** [review]

**Review by AredeSantos — DISMISSED** [review]

**Review by ajpachoriya — DISMISSED** [review]

File: qudini-reactive-graphql/README.md
Comment (Vishal-Rao-SU) [inline qudini-reactive-graphql/README.md:1]: Introspection is taken care on another ticket

File: qudini-reactive-graphql/src/main/java/com/qudini/reactive/graphql/ReactiveGraphQLAutoConfiguration.java [hunk 5]
```diff
@@ -7,8 +7,10 @@
 import com.qudini.reactive.graphql.http.LoggingContextAwareExceptionHandler;
 import com.qudini.reactive.graphql.scalar.Scalar;
 import graphql.execution.DataFetcherExceptionHandler;
+import graphql.schema.GraphQLCodeRegistry;
```
Comment (jmaister) [inline qudini-reactive-graphql/src/main/java/com/qudini/reactive/graphql/ReactiveGraphQLAutoConfiguration.java:10]: Remove unused dependencies.

Comment (Vishal-Rao-SU) [reply qudini-reactive-graphql/src/main/java/com/qudini/reactive/graphql/ReactiveGraphQLAutoConfiguration.java]: taken care.

**Review by jmaister — APPROVED** [review]

**Review by ajpachoriya — APPROVED** [review]

**Review by jmaister — APPROVED** [review]

**Pull request closed without merging**